File: WebDriver-Series/WebDriver.Series.Tests/Controls/KendoGrid.cs
using Newtonsoft.Json;
using OpenQA.Selenium;
using System;
using System.Collections.Generic;
using System.Text;
using WebDriver.Series.Tests.Enums;

namespace WebDriver.Series.Tests.Controls
{
    public class KendoGrid
    {
        private readonly string gridId;
        private readonly IJavaScriptExecutor driver;

        public KendoGrid(IWebDriver driver, IWebElement gridDiv)
        {
            this.gridId = gridDiv.GetAttribute("id");
            this.driver = (IJavaScriptExecutor)driver;
        }

        public void RemoveFilters()
        {
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "grid.dataSource.filter([]);");
            this.driver.ExecuteScript(jsToBeExecuted);
            ////this.Driver.FullWaitUntilReady();
        }

        public int TotalNumberRows()
        {
            ////this.Driver.FullWaitUntilReady();
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "grid.dataSource.total();");
            var jsResult = this.driver.ExecuteScript(jsToBeExecuted);
            return int.Parse(jsResult.ToString());
        }

        public void Reload()
        {
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "grid.dataSource.read();");
            this.driver.ExecuteScript(jsToBeExecuted);
            ////this.Driver.FullWaitUntilReady();
        }

        public int GetPageSize()
        {
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "return grid.dataSource.pageSize();");
            var currentResponse = this.driver.ExecuteScript(jsToBeExecuted);
            int pageSize = int.Parse(currentResponse.ToString());
            return pageSize;
        }

        public void ChangePageSize(int newSize)
        {
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "grid.dataSource.pageSize(", newSize, ");");
            this.driver.ExecuteScript(jsToBeExecuted);
            ////this.Driver.FullWaitUntilReady();
        }

        public void NavigateToPage(int pageNumber)
        {
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "grid.dataSource.page(", pageNumber, ");");
            this.driver.ExecuteScript(jsToBeExecuted);
        }

        public void Sort(string columnName, SortType sortType)
        {
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "grid.dataSource.sort({field: '", columnName, "', dir: '", sortType.ToString().ToLower(), "'});");
            this.driver.ExecuteScript(jsToBeExecuted);
            ////this.Driver.FullWaitUntilReady();
        }

        public List<T> GetItems<T>() where T : class
        {
            ////this.Driver.FullWaitUntilReady();
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "return JSON.stringify(grid.dataSource.data());");
            var jsResults = this.driver.ExecuteScript(jsToBeExecuted);
            var items = JsonConvert.DeserializeObject<List<T>>(jsResults.ToString());
            return items;
        }

        public void Filter(string columnName, FilterOperator filterOperator, string filterValue)
        {
            this.Filter(new GridFilter(columnName, filterOperator, filterValue));
        }

        public void Filter(params GridFilter[] gridFilters)
        {
            string jsToBeExecuted = this.GetGridReference();
            StringBuilder sb = new StringBuilder();
            sb.Append(jsToBeExecuted);
            sb.Append("grid.dataSource.filter({ logic: \"and\", filters: [");
            foreach (var currentFilter in gridFilters)
            {
                DateTime filterDateTime;
                bool isFilterDateTime = DateTime.TryParse(currentFilter.FilterValue, out filterDateTime);
                string filterValueToBeApplied =
                                               isFilterDateTime ? string.Format("new Date({0}, {1}, {2})", filterDateTime.Year, filterDateTime.Month - 1, filterDateTime.Day) :
                                                string.Format("\"{0}\"", currentFilter.FilterValue);
                string kendoFilterOperator = this.ConvertFilterOperatorToKendoOperator(currentFilter.FilterOperator);
                sb.Append(string.Concat("{ field: \"", currentFilter.ColumnName, "\", operator: \"", kendoFilterOperator, "\", value: ", filterValueToBeApplied, " },"));
            }
            sb.Append("] });");
            jsToBeExecuted = sb.ToString().Replace(",]", "]");
            this.driver.ExecuteScript(jsToBeExecuted);
            ////this.Driver.FullWaitUntilReady();
        }

        public int GetCurrentPageNumber()
        {
            string jsToBeExecuted = this.GetGridReference();
            jsToBeExecuted = string.Concat(jsToBeExecuted, "return grid.dataSource.page();");
            var result = this.driver.ExecuteScript(jsToBeExecuted);
            int pageNumber = int.Parse(result.ToString());
            return pageNumber;
        }

        private string GetGridReference()
        {
            string initializeKendoGrid = string.Format("var grid = $('#{0}').data('kendoGrid');", this.gridId);

            return initializeKendoGrid;
        }

        private string ConvertFilterOperatorToKendoOperator(FilterOperator filterOperator)
        {
            string kendoFilterOperator = string.Empty;
            switch (filterOperator)
            {
                case FilterOperator.EqualTo:
                    kendoFilterOperator = "eq";
                    break;
                case FilterOperator.NotEqualTo:
                    kendoFilterOperator = "neq";
                    break;
                case FilterOperator.LessThan:
                    kendoFilterOperator = "lt";
                    break;
                case FilterOperator.LessThanOrEqualTo:
                    kendoFilterOperator = "lte";
                    break;
                case FilterOperator.GreaterThan:
                    kendoFilterOperator = "gt";
                    break;
                case FilterOperator.GreaterThanOrEqualTo:
                    kendoFilterOperator = "gte";
                    break;
                case FilterOperator.StartsWith:
                    kendoFilterOperator = "startswith";
                    break;
                case FilterOperator.EndsWith:
                    kendoFilterOperator = "endswith";
                    break;
                case FilterOperator.Contains:
                    kendoFilterOperator = "contains";
                    break;
                case FilterOperator.NotContains:
                    kendoFilterOperator = "doesnotcontain";
                    break;
                case FilterOperator.IsAfter:
                    kendoFilterOperator = "gt";
                    break;
                case FilterOperator.IsAfterOrEqualTo:
                    kendoFilterOperator = "gte";
                    break;
                case FilterOperator.IsBefore:
                    kendoFilterOperator = "lt";
                    break;
                case FilterOperator.IsBeforeOrEqualTo:
                    kendoFilterOperator = "lte";
                    break;
                default:
                    throw new ArgumentException("The specified filter operator is not supported.");
            }

            return kendoFilterOperator;
        }
    }
}

File: java/WebAutomation-Series/src/test/java/kendogrid/components/KendoGrid.java
package kendogrid.components;

import com.google.gson.Gson;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;

public class KendoGrid {
    private final String _gridId;
    private final JavascriptExecutor driver;
    private final WebDriverWait _wait;

    public KendoGrid(WebDriver driver, WebElement gridDiv) {
        _gridId = gridDiv.getAttribute("id");
        this.driver = (JavascriptExecutor) driver;
        _wait = new WebDriverWait(driver, 30);
    }

    public void removeFilters() {
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "grid.dataSource.filter([]);";
        driver.executeScript(jsToBeExecuted);
        waitForAjax();
    }

    public int totalNumberRows() {
        waitForAjax();
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "grid.dataSource.total();";
        var jsResult = driver.executeScript(jsToBeExecuted);
        return Integer.parseInt(jsResult.toString());
    }

    public void reload() {
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "grid.dataSource.read();";
        driver.executeScript(jsToBeExecuted);
        waitForAjax();
    }

    public int getPageSize() {
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "return grid.dataSource.pageSize();";
        var currentResponse = driver.executeScript(jsToBeExecuted);
        var pageSize = Integer.parseInt(currentResponse.toString());
        return pageSize;
    }

    public void changePageSize(int newSize) {
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "grid.dataSource.pageSize(" + newSize + ");";
        driver.executeScript(jsToBeExecuted);
        waitForAjax();
    }

    public void navigateToPage(int pageNumber) {
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "grid.dataSource.page(" + pageNumber + ");";
        driver.executeScript(jsToBeExecuted);
    }

    public void sort(String columnName, SortType sortType) {
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "grid.dataSource.sort({field: '" + columnName + "', dir: '" + sortType.toString().toLowerCase() + "'});";
        driver.executeScript(jsToBeExecuted);
        waitForAjax();
    }

    public <T> List<T> getItems() {
        waitForAjax();
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "return JSON.stringify(grid.dataItems());";
        var jsResults = driver.executeScript(jsToBeExecuted);

        Gson gson = new Gson();
        List<T> items = gson.fromJson(jsResults.toString(), ArrayList.class);
        return items;
    }

    public void filter(String columnName, FilterOperator filterOperator, String filterValue) throws Exception {
        filter(new GridFilter(columnName, filterOperator, filterValue));
    }

    public void filter(GridFilter... gridFilters) throws Exception {
        var jsToBeExecuted = getGridReference();
        var sb = new StringBuilder();
        sb.append(jsToBeExecuted);
        sb.append("grid.dataSource.filter({ logic: \"and\", filters: [");
        for (var currentFilter: gridFilters) {
            var filterValueToBeApplied =  String.format("\"%s\"", currentFilter.getFilterValue());
            try {
                LocalDateTime filterDateTime = LocalDateTime.parse(currentFilter.getFilterValue());
                filterValueToBeApplied = String.format("new Date(%s, %s, %s})", filterDateTime.getYear(), filterDateTime.getMonthValue() - 1, filterDateTime.getDayOfYear());
            } catch (DateTimeParseException ex) {
                // ignore
            }

            var kendoFilterOperator = convertFilterOperatorToKendoOperator(currentFilter.getFilterOperator());
            sb.append("{ field: \"" + currentFilter.getColumnName() + "\", operator: \"" + kendoFilterOperator + "\", value: " + filterValueToBeApplied + " },");
        }
        sb.append("] });");
        jsToBeExecuted = sb.toString().replace(",]", "]");
        driver.executeScript(jsToBeExecuted);
        waitForAjax();
    }

    public int getCurrentPageNumber() {
        var jsToBeExecuted = getGridReference();
        jsToBeExecuted = jsToBeExecuted + "return grid.dataSource.page();";
        var result = driver.executeScript(jsToBeExecuted);
        var pageNumber = Integer.parseInt(result.toString());
        return pageNumber;
    }

    private String getGridReference() {
        var initializeKendoGrid = String.format("var grid = $('#%s').data('kendoGrid');", _gridId);

        return initializeKendoGrid;
    }

    private String convertFilterOperatorToKendoOperator(FilterOperator filterOperator) throws Exception {
        var kendoFilterOperator = "";
        switch (filterOperator)
        {
            case EqualTo:
                kendoFilterOperator = "eq";
                break;
            case NotEqualTo:
                kendoFilterOperator = "neq";
                break;
            case LessThan:
                kendoFilterOperator = "lt";
                break;
            case LessThanOrEqualTo:
                kendoFilterOperator = "lte";
                break;
            case GreaterThan:
                kendoFilterOperator = "gt";
                break;
            case GreaterThanOrEqualTo:
                kendoFilterOperator = "gte";
                break;
            case StartsWith:
                kendoFilterOperator = "startswith";
                break;
            case EndsWith:
                kendoFilterOperator = "endswith";
                break;
            case Contains:
                kendoFilterOperator = "contains";
                break;
            case NotContains:
                kendoFilterOperator = "doesnotcontain";
                break;
            case IsAfter:
                kendoFilterOperator = "gt";
                break;
            case IsAfterOrEqualTo:
                kendoFilterOperator = "gte";
                break;
            case IsBefore:
                kendoFilterOperator = "lt";
                break;
            case IsBeforeOrEqualTo:
                kendoFilterOperator = "lte";
                break;
            default:
                throw new Exception("The specified filter operator is not supported.");
        }

        return kendoFilterOperator;
    }

    private void waitForAjax()
    {
        _wait.until(d -> (Boolean)((JavascriptExecutor)d).executeScript("return jQuery.active == 0"));
    }
}
